text/0000-trinary-trytes.md

```diff
@@ -0,0 +1,180 @@
++ Feature Name: `trinary-trytes`
++ Start Date: 2019-09-09
++ RFC PR: [iotaledger/bee-rfcs#0000](https://github.com/iotaledger/bee-rfcs/pull/4)
++ Bee issue: [iotaledger/bee#0000](https://github.com/iotaledger/bee/issues/50)
+
+# Summary
+
+Define type `Trytes` for higher level usage. `Trytes` is going to be similar to
+`String` which is a vector of `u8` under the structure. It provides almost same
+methods but different signatures compares to `String`, because we are going to
+make sure every inputs to `Trytes` are valid **tryte alphabete**. **Tryte alphabete**
+is a table to make trytes easy to read. Since it serves this purpose and would
+interact with underlying ternary operations in most cases, `Trytes` will behave
+like string literal and each charaters are actually UTF-8 encoded character. So
+`9` has value of 57 and `A` to `Z` have values of `65` to `90` in its vector.
+
+# Motivation
+
+The reason to implement trytes only at the moment is not only to be loosly
+coupled from **trits**, but also layout what a ternary type could looks like. We
+will eventually have various encoding techniques. It may become a mess if we
+implement all in one structure, or seperate files but without a well-defined layout.
+For the reason to use vector of `u8` is because it's sufficent in most cases,
+implementations of `trit` encoding would need `bytes` from tokio or maybe
+`bitflags` instead. I would like to make it be able to represent as string
+literals and slice of values at the same time. As for choice of tryte alphabete
+being UTF-8 encoded is because we expect most users would like to conver strings
+and characters to `Trytes`. It makes more scene to stay same encoding as others
+in Rust. Maybe we will need to convert byte literals in the end, but still this
+can be done afterwards.
+
+# Detailed design
+
+## Trytes
+
+Type `Trytes` is a vector of valus `Vec<u8>` and moreover a smart pointer.
+The structure simply looks like this:
+
+```rust
+pub type Tryte = u8;
```

Comment:
Using a byte per tryte is certainly the most intuitive solution. @Alex6323 and I had an alternative idea that would increase speed and reduce memory: encode 3 trytes with 2 bytes:
* this reduces memory requirements by 33%
* we can compare vectors faster as we compare less bytes
* we do not need to compress and decompress trytes when receiving/sending transaction to neighbors but can simply forward the internal representation

Reply:
You are talking trits encoding which we will definitely have various encoding implementations like TE27 used in troika. I would like to put encoding and operation logic to trits only. So trytes can be easy to understand and serves as tryte alphabet. We still need to convert between different encodings afterwards though.

Reply:
What about `Trytes::from_human_readable("SOME9TRYTE9SEQUENCE");`? Internal representation could still be compressed to 9 trits per 2 bytes to reduce memory requirements and increase performance (as no compression/decompression is required for transmission to neighbors).

Reply:
This implementation is for human readable strings. All encoding technique I encourage stay in any self-defined trit data structure (that say you have a trit type called `9TritsPer2Bytes`). I think the internal of bee would no need for `Trytes`. Unless you want to repressent or debug, in this case you can define a method of your `9TritsPer2Bytes` to convert to `Trytes`.

Reply:
Ah, perfect. But do we actually need a type just for human readable strings? Wouldn't `CompactTrinary::to_human_readable(&self) -> String` suffice, where `CompactTrinary` is the type that internally stores the 9 trit / 2 byte encoding?

Reply:
Yes, it's defined as `String` in `iota-lib-rs`. But since we are going to re-implement with TDD, I'm thinking to define a more secure one. The methods I proposed have almost same signatures with string but many are return with `Result` type to make sure it's valid input. But indeed some might be overkill, I'm going to reduce some with zesterer's advice.